### ISSUE TITLE
🐛 Multiple false positives reported by biome check

### ISSUE THREAD

**Issue by tsatsujnr139** [biomejs/biome#9478]
### Environment information

<details>

```bash

CLI:
  Version:                      2.4.7
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.13.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.32.1

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.jsonc
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true
  HTML full support enabled:    unset

Workspace:
  Open Documents:               0
```

</details>

### What happened?

1. Run `biome check`
2. Biome incorrectly reports issues mainly related to the 2 rules below. This used to work correctly before, but now even downgrading to my previous version doesn't resolve the issue.
 - `lint/correctness/noUndeclaredDependencies`
 - `lint/correctness/noUnresolvedImports`

eg. 
```bash
tooling/turborepo-remote-cache/src/middleware/rate-limiter.ts:1:39 lint/correctness/noUndeclaredDependencies ━━━━━━━━━━

  ✖ Dependency @hono-rate-limiter/cloudflare isn't specified in tooling/turborepo-remote-cache/package.json.
  
  > 1 │ import { cloudflareRateLimiter } from "@hono-rate-limiter/cloudflare";
      │                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    2 │ import type { Context, Next } from "hono";
```

### Expected result

It should not report the issues incorrectly.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by Gint367** [comment]
got same problem, the problem arise when I update from version 2.4.6 to 2.4.7
the time needed to run the check is also a lot worse on 2.4.7. on the top row is the 2.4.7, the bottom one is the downgraded 2.4.6

<img width="627" height="312" alt="Image" src="https://github.com/user-attachments/assets/9a04dc25-06a8-4f2e-a5c3-09b997eac033" />

**Comment by dyc3** [comment]
@Gint367 the performance thing seems like a different issue. Can you file a new issue **with a reproduction** so that we can troubleshoot it?

**Comment by tsatsujnr139** [comment]
The repro is not applicable here because it's related to reporting false positives in libraries like react and useEffect imports.

The playground isn't installing actual dependencies to detect or checking module resolution.

I can't share a link to a repo either as well as code base is private.

**Comment by ematipico** [comment]
@tsatsujnr139 

As the bot suggested, create a repository (minimal) so you can re create the issue. We also offer a quick cli command to make things fast

**Comment by tsatsujnr139** [comment]
Not been able to successfully reproduce this in a fresh repo, I will keep digging and update here on any findings. Might be something specific to my environment or configuration.

**Comment by naveedpeerzade** [comment]
i had same issue since y'day. but apparently the problem appears to be with ultracite and not biome.

**Comment by mecaota** [comment]
I am also experiencing the same symptoms and am quite troubled by them.

Fortunately, I have a public repository on hand that exhibits these symptoms, so I will submit this as a reproduction repository.

- https://github.com/mecaota/template-mecaota-frontend/pull/167

I'm not sure if I should mention this here, but if you set path aliases in tsconfig using tsconfig-paths, noUndeclaredDependencies will similarly detect them, so I'll also post a repository to reproduce that issue.

- https://github.com/mecaota/template-mecaota-frontend/pull/168

**Comment by ematipico** [comment]
@mecaota a public repository doesn't cut it. We want a **minimal** reproduction.

**Comment by mecaota** [comment]
Since I had some extra time, I created a minimal reproduction code.

The issue is still reproducible in the latest version, 2.4.16. It requires React and TypeScript, and due to an error related to file import rules, I'm providing it as a minimal reproduction repository.

https://github.com/mecaota/biome-repro20260611



- The error rule being discussed in this issue is noUnresolvedImports. The actual React's JavaScript code is exported in commonjs format, but it doesn't appear to be exported in the type definition file. There seems to be an inconsistency here.

- noUndeclaredDependencies seems unrelated to the issue raised in this issue. However, because it ignores the paths in tsconfig.json, an error occurs in this minimal reproduction repository. Some kind of option or change is likely needed, and I intend to create a separate issue for that.

**Comment by mecaota** [comment]
@ematipico If you don't mind, could you please check out this minimal configuration repository?

**Comment by siketyan** [comment]
@mecaota 

Thank you for creating a repro, I have checked it and confirmed two issues are there.
However, I believe the two issues are coming from different causes.

a) For the `noUndeclaredDependencies` one, it seems Biome currently has a limitation that it can't distinguish between packages and path aliases of the `@foo/...` format. As a workaround, you can use `@/foo/...` style that is distinguishable.

b) For the `noUnresolvedImports` one, it seems Biome isn't reading the type definition file correctly. It **is** declared in the file [^1], but in a namespace. Maybe Biome doesn't resolve a symbol in a namespace correctly.

[^1]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a166beb3b5889cd396ddca574e7305657c3de844/types/react/index.d.ts#L782

Can you file new issues **for each** issue so we can track them individually?

**Comment by mecaota** [comment]
@siketyan 

Ok, published these issues.

a) noUndeclaredDependencies -> https://github.com/biomejs/biome/issues/10607

b) noUnresolvedImports -> https://github.com/biomejs/biome/issues/10609

**Comment by tsatsujnr139** [comment]
I was able to figure out the issue on my end was due to ultracite biome configuration issues and not directly related to biome.
